### DISCUSSION TITLE
About Sending ACK

### DISCUSSION THREAD

**Discussion by tolgaeng** [jgromes/RadioLib#329] [Q&A]
Hi everyone!

I'm trying to send a small picture (about 1KB) with LoRa. I turned the picture into small packets about 200 bayt and want to send it with that way. To do that I'll need a ACK mechanism. Basically I'll send a packet to receiver and receiver will send me the packet number as ACK. If I receive ACK from transmitter then I'll send the next packet. It'll go all along the packets.

The part that I stuck is sending ACK from receiver. I can send ACK for just one time if I receive the packet without error. But just one sending process is not enough to work. Need to send continuously old ACK data. So transmitter can catch that easily and jump to the next packet.

I'm not sure that I explain correctly but If you ask a part I'll try to explain it much better.

Here's my simple version of my code:

    byte packetData[packetSize + 1];
    int state = radio.receive(packetData, packetSize + 1);

    if (state == ERR_NONE) {

    packetNumber = packetData[0];

    delay(1000);

    radio.sleep();
    radio.begin();
  
     if (packetsReceived == false) {

      Serial.write(packetData + 1, packetSize);

      if (packetData[0] == lastCharacter) {
          packetsReceived = true;
        }
      }

    radio.transmit(packetNumber, 1, 0);

    radio.sleep();
    radio.begin();
    }

    radio.sleep();
    radio.begin();
    radio.transmit(packetNumber, 1, 0);
    radio.sleep();
    radio.begin();

    delay(10);
 
    radio.sleep();

It's not working correctly. There are a lot of radio.sleep or begin functions, I'm actually not sure that is I need them but I worked better then without them. And there's a weird things like sometimes when I change the code Arduino waits like 7 seconds however there's no any delay function. I think that is listening time or something but I cannot find it in the libraries.

Thank you.

**Comment by jgromes** [comment]
In principle the idea is solid, however, I would not choose blocking methods `radio.transmit()`/`radio.receive()` to implment it. At the moment, you're only receiving while within `radio.receive()`. Take a look at the Transmit_Interrupt/Receive_Interrupt/SX126x_PingPong examples.

Here's an outline of what I would do:

1. Keep both ends of the link in listen mode.
2. Send the first image packet and after that switch back to listen mode.
3. When the "image receiver" gets an image packet, transmit the ACK and then switch back to receive mode.
4. When the "image transmitter" gets the ACK, goto 2.

Also, don't use `radio.begin()` to wake up from sleep - it will do a complete reset and initialization. `radio.begin()` should only be used for power-up init or when switching modems. Use `radio.standby()` to go to standby from sleep, or `radio.startReceive()` to go directly to listen mode.
